app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt

```diff
@@ -0,0 +1,121 @@
+
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.graphics.Paint
+import android.graphics.drawable.BitmapDrawable
+import android.graphics.drawable.Drawable
+import android.text.SpannableStringBuilder
+import android.text.Spanned
+import android.text.SpannedString
+import android.text.style.ReplacementSpan
+import android.view.View
+
+import com.bumptech.glide.Glide
+import com.bumptech.glide.request.target.CustomTarget
+import com.bumptech.glide.request.target.Target
+import com.bumptech.glide.request.transition.Transition
+import com.keylesspalace.tusky.entity.Emoji
+
+import java.lang.ref.WeakReference
+import java.util.regex.Pattern
+import androidx.preference.PreferenceManager
+
+/**
+ * replaces emoji shortcodes in a text with EmojiSpans
+ * @param text the text containing custom emojis
+ * @param emojis a list of the custom emojis (nullable for backward compatibility with old mastodon instances)
+ * @param view a reference to the a view the emojis will be shown in (should be the TextView, but parents of the TextView are also acceptable)
+ * @return the text with the shortcodes replaced by EmojiSpans
+*/
+fun emojifyText(text: Spanned, emojis: List<Emoji>?, view: View) : Spanned {
+    if (emojis != null && emojis.isNotEmpty()) {
+        val builder = SpannableStringBuilder(text)
+        for (emoji in emojis) {
+            val pattern = StringBuilder(":").append(emoji.shortcode).append(":")
+            val matcher = Pattern.compile(pattern.toString(), Pattern.LITERAL)
+                .matcher(text)
+            while(matcher.find()) {
+                val span = EmojiSpan(WeakReference<View>(view))
+                builder.setSpan(span, matcher.start(), matcher.end(), 0);
+                Glide.with(view)
+                    .asBitmap()
+                    .load(emoji.url)
+                    .into(span.getTarget())
+            }
+        }
+        
+        return builder
+    }
+    
+    return text
+}
+
+fun emojifyString(string: String, emojis: List<Emoji>?, view: View) : Spanned {
+    return emojifyText(SpannedString(string), emojis, view)
+}
+
+public class EmojiSpan(val viewWeakReference: WeakReference<View>) : ReplacementSpan() {
+    var imageDrawable: Drawable? = null
+    
+    override fun getSize(paint: Paint, text: CharSequence, start: Int, end: Int, fm: Paint.FontMetricsInt?) : Int {
+        if (fm != null) {
+            /* update FontMetricsInt or otherwise span does not get drawn when
+             * it covers the whole text */
+            val metrics = paint.fontMetricsInt
+            fm.top = metrics.top
+            fm.ascent = metrics.ascent
+            fm.descent = metrics.descent
+            fm.bottom = metrics.bottom
+        }
+        
+        return (paint.textSize * 1.2).toInt()
+    }
+    
+    override fun draw(canvas: Canvas, text: CharSequence, start: Int, end: Int, x: Float, top: Int, y: Int, bottom: Int, paint: Paint) {
+        if (imageDrawable == null)
```

Comment:
Sorry for long answer. 
I'm kinda new to Kotlin, but it seems that if we have imageDrawable declared as `var`, it may be changed from other thread(theoretically?) so if I move the code under ?.let block, I still can't remove !!.

I don't see a way how I can to improve that other than doing `imageDrawable?.let { drawable -> ... }`

Reply:
yes, inside the `let` block you can't use `imageDrawable` because its still a `var`, so either use the implicit `it` or give it another name like you proposed